src/devices/imuBosch_BNO055/imuBosch_BNO055.cpp

```diff
@@ -463,93 +462,101 @@ void BoschIMU::run()
-    {
+    if (sendReadCommand(REG_MAGN_DATA, 6, response, "Read Magnetometer")) {
         // Manually compose the data to safely handling endianess
-        raw_data[0] = response[3] << 8 | response[2];
-        raw_data[1] = response[5] << 8 | response[4];
-        raw_data[2] = response[7] << 8 | response[6];
-        data[ 9] = (double) raw_data[0]/16.0;
-        data[10] = (double) raw_data[1]/16.0;
-        data[11] = (double) raw_data[2]/16.0;
-    //     yDebug() << "Magn x: " << data[9] << "y: " << data[10] << "z: " << data[11];
+        raw_data[0]  = response[3] << 8 | response[2];
+        raw_data[1]  = response[5] << 8 | response[4];
+        raw_data[2]  = response[7] << 8 | response[6];
+        data_tmp[9]  = (double)raw_data[0] / 16.0;
+        data_tmp[10] = (double)raw_data[1] / 16.0;
+        data_tmp[11] = (double)raw_data[2] / 16.0;
+        // yDebug() << "Magn x: " << data[9] << "y: " << data[10] << "z: " << data[11];
     }
-    else
+    else {
         errs.magnError++;
+    }
 
     ///////////////////////////////////////////
     //
     //      Read Quaternion
     //
     ///////////////////////////////////////////
 
-    if(sendReadCommand(REG_QUATERN_DATA, 8, response, "Read quaternion") )
-    {
+    yarp::math::Quaternion quaternion_tmp;
+    if (sendReadCommand(REG_QUATERN_DATA, 8, response, "Read quaternion")) {
         // Manually compose the data to safely handling endianess
         raw_data[0] = response[3] << 8 | response[2];
         raw_data[1] = response[5] << 8 | response[4];
         raw_data[2] = response[7] << 8 | response[6];
         raw_data[3] = response[9] << 8 | response[8];
 
-        quaternion.w() = ((double) raw_data[0])/(2<<13);
-        quaternion.x() = ((double) raw_data[1])/(2<<13);
-        quaternion.y() = ((double) raw_data[2])/(2<<13);
-        quaternion.z() = ((double) raw_data[3])/(2<<13);
+        quaternion_tmp.w() = ((double)raw_data[0]) / (2 << 13);
+        quaternion_tmp.x() = ((double)raw_data[1]) / (2 << 13);
+        quaternion_tmp.y() = ((double)raw_data[2]) / (2 << 13);
+        quaternion_tmp.z() = ((double)raw_data[3]) / (2 << 13);
 
         RPY_angle.resize(3);
-        RPY_angle = yarp::math::dcm2rpy(quaternion.toRotationMatrix());
-        data[0] = RPY_angle[0] * 180/M_PI;
-        data[1] = RPY_angle[1] * 180/M_PI;
-        data[2] = RPY_angle[2] * 180/M_PI;
+        RPY_angle   = yarp::math::dcm2rpy(quaternion.toRotationMatrix());
+        data_tmp[0] = RPY_angle[0] * 180 / M_PI;
+        data_tmp[1] = RPY_angle[1] * 180 / M_PI;
+        data_tmp[2] = RPY_angle[2] * 180 / M_PI;
     }
-    else
+    else {
         errs.quatError++;
+    }
 
     // If 100ms have passed since the last received message
     if (timeStamp+0.1 < yarp::os::Time::now())
     {
 //         status=TIMEOUT;
     }
-    mutex.post();
+
+    // Protect only this section in order to avoid slow race conditions when gathering this data
+    mutex.lock();
```

Comment:
You are right, I just fear some compiler optimization on some OS may remove the scope since there is no real reason for it ...
It'd be very weird behaviour for a compiler, but I'll sleep better with the old fashioned lock()  :smile: 
Anyway it is fine if you are sure it does what we want it to